src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -841,5 +851,31 @@ object PruneDeadFields {
       }
     }
   }
+
+  def upcast(ir: MatrixIR, rType: MatrixType,
+    upcastRows: Boolean = true,
+    upcastCols: Boolean = true,
+    upcastGlobals: Boolean = true,
+    upcastEntries: Boolean = true): MatrixIR = {
+
+    if (ir.typ == rType || !(upcastRows || upcastCols || upcastGlobals || upcastEntries))
+      ir
+    else {
+      var mt = ir
+      if (upcastEntries && mt.typ.entryType != rType.entryType)
+        mt = MatrixMapEntries(mt, SelectFields(Ref("g", mt.typ.entryType), rType.entryType.fieldNames))
```

Comment:
this won't work -- need to support the full nested type. Just use the existing `upcast` for value IRs.

Reply:
ah, good call.